File: android/app/src/main/java/com/soar/tracker/ui/components/SensorDisplay.kt
package com.soar.tracker.ui.components

import java.util.Locale
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlin.math.sqrt

@Composable
fun SensorDisplay(
    accelX: Double?,
    accelY: Double?,
    accelZ: Double?,
    pressureHpa: Double?,
    verticalSpeedMps: Double?,
    latitude: Double?,
    longitude: Double?,
    altitudeMeters: Double?,
    altitudeAglFeet: Int?,
    gpsBearingDegrees: Double?,
    magneticHeadingDegrees: Double?,
    magneticDeclinationDegrees: Double?,
    groundPressureHpa: Double?,
    onSetAltimeter: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant,
        ),
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text("Sensors", style = MaterialTheme.typography.titleMedium)
            Spacer(modifier = Modifier.height(8.dp))

            // Row 1: G-Force, Vario, Pressure
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                val gForce = if (accelX != null && accelY != null && accelZ != null) {
                    sqrt(accelX * accelX + accelY * accelY + accelZ * accelZ)
                } else {
                    null
                }

                SensorValue(
                    label = "G-Force",
                    value = gForce?.let { String.format(Locale.US, "%.1f G", it) } ?: "--",
                )

                val varioFpm = verticalSpeedMps?.let { it * 196.85 }
                SensorValue(
                    label = "Vario",
                    value = varioFpm?.let {
                        val sign = if (it >= 0) "+" else ""
                        String.format(Locale.US, "%s%.0f fpm", sign, it)
                    } ?: "--",
                )

                // Pressure in both hPa and inHg
                val pressureText = if (pressureHpa != null) {
                    val inHg = pressureHpa * 0.02953
                    String.format(Locale.US, "%.1f hPa\n%.2f inHg", pressureHpa, inHg)
                } else {
                    "--"
                }
                SensorValue(
                    label = "Pressure",
                    value = pressureText,
                )
            }

            Spacer(modifier = Modifier.height(12.dp))

            // Row 2: GNSS Alt, AGL, Position
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                val altFt = altitudeMeters?.let { it * 3.28084 }
                SensorValue(
                    label = "GNSS Alt",
                    value = altFt?.let { String.format(Locale.US, "%.0f ft", it) } ?: "--",
                )

                SensorValue(
                    label = "AGL",
                    value = altitudeAglFeet?.let { String.format(Locale.US, "%d ft", it) } ?: "--",
                )

                SensorValue(
                    label = "Position",
                    value = if (latitude != null && longitude != null) {
                        String.format(Locale.US, "%.4f\n%.4f", latitude, longitude)
                    } else {
                        "--"
                    },
                )
            }

            // Row 3: Heading Magnetic (compass), Heading True (computed), Declination
            Spacer(modifier = Modifier.height(12.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                SensorValue(
                    label = "Hdg (Mag)",
                    value = magneticHeadingDegrees?.let {
                        String.format(Locale.US, "%.0f\u00B0", it)
                    } ?: "--",
                )

                // True heading = magnetic heading + declination
                val trueHeading = if (magneticHeadingDegrees != null && magneticDeclinationDegrees != null) {
                    var hdg = magneticHeadingDegrees + magneticDeclinationDegrees
                    if (hdg < 0) hdg += 360.0
                    if (hdg >= 360) hdg -= 360.0
                    hdg
                } else {
                    null
                }
                SensorValue(
                    label = "Hdg (True)",
                    value = trueHeading?.let {
                        String.format(Locale.US, "%.0f\u00B0", it)
                    } ?: "--",
                )

                SensorValue(
                    label = "Track (GPS)",
                    value = gpsBearingDegrees?.let {
                        String.format(Locale.US, "%.0f\u00B0", it)
                    } ?: "--",
                )
            }

            // Row 4: Declination
            Spacer(modifier = Modifier.height(12.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                SensorValue(
                    label = "Declination",
                    value = magneticDeclinationDegrees?.let {
                        val sign = if (it >= 0) "E" else "W"
                        String.format(Locale.US, "%.1f\u00B0 %s", Math.abs(it), sign)
                    } ?: "--",
                )
            }

            Spacer(modifier = Modifier.height(12.dp))
            HorizontalDivider()
            Spacer(modifier = Modifier.height(12.dp))

            // Altimeter section
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // Altimeter reading
                val altimeterFt = if (pressureHpa != null && groundPressureHpa != null && groundPressureHpa > 0) {
                    145366.45 * (1 - Math.pow(pressureHpa / groundPressureHpa, 0.190284))
                } else {
                    null
                }
                SensorValue(
                    label = "Altimeter",
                    value = altimeterFt?.let { String.format(Locale.US, "%.0f ft", it) } ?: "--",
                )

                // Altimeter setting (the reference pressure)
                SensorValue(
                    label = "Altimeter Setting",
                    value = groundPressureHpa?.let {
                        val inHg = it * 0.02953
                        String.format(Locale.US, "%.2f inHg\n%.1f hPa", inHg, it)
                    } ?: "Not set",
                )

                // Set button
                TextButton(
                    onClick = onSetAltimeter,
                    enabled = pressureHpa != null,
                    contentPadding = ButtonDefaults.TextButtonContentPadding,
                ) {
                    Text("Set")
                }
            }
        }
    }
}

@Composable
private fun SensorValue(label: String, value: String) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            text = value,
            style = MaterialTheme.typography.titleMedium,
        )
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

File: ios/SOARTracker/UI/Components/SensorDisplay.swift
import SwiftUI

struct SensorDisplay: View {
    let sensorData: SensorCollector.SensorData
    let response: TrackerFixResponse?
    let magneticHeading: Double?
    let groundPressureHpa: Double?
    let gnssAltitudeMeters: Double?
    let latitude: Double?
    let longitude: Double?
    let gpsTrack: Double?
    let verticalSpeedMps: Double?
    let onSetAltimeter: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            // Header
            Text("Sensors")
                .font(.headline)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 16)
                .padding(.top, 12)
                .padding(.bottom, 8)

            // Row 1: G-Force, Vario, Pressure
            HStack(spacing: 0) {
                SensorCell(label: "G-Force", value: gForceText)
                SensorCell(label: "Vario", value: varioText, color: varioColor)
                SensorCell(label: "Pressure", value: pressureText)
            }

            Divider().padding(.horizontal, 16)

            // Row 2: GNSS Alt, AGL, Position
            HStack(spacing: 0) {
                SensorCell(label: "GNSS Alt", value: gnssAltText)
                SensorCell(label: "AGL", value: aglText)
                SensorCell(label: "Position", value: positionText, isSmall: true)
            }

            Divider().padding(.horizontal, 16)

            // Row 3: Mag Heading, True Heading, GPS Track
            HStack(spacing: 0) {
                SensorCell(label: "Mag Hdg", value: magHeadingText)
                SensorCell(label: "True Hdg", value: trueHeadingText)
                SensorCell(label: "GPS Track", value: gpsTrackText)
            }

            Divider().padding(.horizontal, 16)

            // Altimeter section
            altimeterSection
        }
        .background(Color.cardBackground)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    // MARK: - Computed Values

    private var gForceText: String {
        guard let x = sensorData.accelX, let y = sensorData.accelY, let z = sensorData.accelZ else {
            return "--"
        }
        let magnitude = sqrt(x * x + y * y + z * z) / 9.81
        return String(format: "%.2f g", magnitude)
    }

    private var varioFpm: Double? {
        guard let vs = verticalSpeedMps else { return nil }
        // m/s to ft/min: multiply by 196.85
        return vs * 196.85
    }

    private var varioText: String {
        guard let fpm = varioFpm else { return "--" }
        let sign = fpm >= 0 ? "+" : ""
        return "\(sign)\(Int(fpm)) fpm"
    }

    private var varioColor: Color {
        guard let fpm = varioFpm else { return .primary }
        if fpm > 50 { return .soarGreen }
        if fpm < -50 { return .soarRed }
        return .primary
    }

    private var pressureText: String {
        guard let hPa = sensorData.pressureHpa else { return "--" }
        return "\(Formatters.inHg(hPa: hPa)) inHg"
    }

    private var gnssAltText: String {
        guard let alt = gnssAltitudeMeters else { return "--" }
        return "\(Int(Formatters.metersToFeet(alt))) ft"
    }

    private var aglText: String {
        guard let agl = response?.altitudeAglFeet else { return "--" }
        return "\(Int(agl)) ft"
    }

    private var positionText: String {
        guard let lat = latitude, let lon = longitude else { return "--" }
        let latStr = String(format: "%.4f\u{00B0}%@", abs(lat), Formatters.latCardinal(lat))
        let lonStr = String(format: "%.4f\u{00B0}%@", abs(lon), Formatters.lonCardinal(lon))
        return "\(latStr) \(lonStr)"
    }

    private var magHeadingText: String {
        guard let heading = magneticHeading else { return "--" }
        return "\(Int(heading))\u{00B0}"
    }

    private var trueHeadingText: String {
        guard let mag = magneticHeading, let dec = response?.magneticDeclinationDegrees else { return "--" }
        let trueHdg = (mag + dec + 360).truncatingRemainder(dividingBy: 360)
        return "\(Int(trueHdg))\u{00B0}"
    }

    private var gpsTrackText: String {
        guard let track = gpsTrack else { return "--" }
        return "\(Int(track))\u{00B0}"
    }

    // MARK: - Altimeter Section

    @ViewBuilder
    private var altimeterSection: some View {
        VStack(spacing: 8) {
            HStack {
                Text("Altimeter")
                    .font(.subheadline.bold())

                Spacer()

                if let groundP = groundPressureHpa {
                    Text(String(format: "Ref: %.1f hPa", groundP))
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }

                Button("Set", action: onSetAltimeter)
                    .font(.caption)
                    .buttonStyle(.bordered)
                    .controlSize(.mini)
                    .disabled(sensorData.pressureHpa == nil)
            }

            if let currentP = sensorData.pressureHpa, let groundP = groundPressureHpa {
                let alt = Formatters.barometricAltitude(pressureHpa: currentP, groundPressureHpa: groundP)
                Text("\(Int(alt)) ft")
                    .font(.title2.monospacedDigit())
                    .fontWeight(.semibold)
            } else {
                Text("--")
                    .font(.title2)
                    .foregroundStyle(.secondary)
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }
}

// MARK: - Sensor Cell

private struct SensorCell: View {
    let label: String
    let value: String
    var color: Color = .primary
    var isSmall: Bool = false

    var body: some View {
        VStack(spacing: 2) {
            Text(label)
                .font(.caption2)
                .foregroundStyle(.secondary)
            Text(value)
                .font(isSmall ? .caption.monospacedDigit() : .callout.monospacedDigit())
                .fontWeight(.medium)
                .foregroundStyle(color)
                .lineLimit(1)
                .minimumScaleFactor(0.7)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 8)
    }
}
